### PR TITLE
Use Docker multi-stage builds to reduce final docker image size

### PR DESCRIPTION
**Pull request by prymitive**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -56,6 +56,9 @@ bindata_assetfs.go: .build/deps.ok .build/bindata_assetfs.$(GO_BINDATA_MODE) $(A
 $(NAME): .build/deps.ok .build/vendor.ok bindata_assetfs.go $(SOURCES)
 	go build -ldflags "-X main.version=$(VERSION)"
 
+$(NAME)-static: .build/deps.ok .build/vendor.ok bindata_assetfs.go $(SOURCES)
```
Comment (mattbostock) [inline Makefile]: What's this change? Should it be in a separate commit?

File: Dockerfile [hunk 2]
```diff
@@ -1,16 +1,11 @@
...
+RUN make -C /go/src/github.com/cloudflare/unsee VERSION="${VERSION:-dev}" unsee-static
 
-RUN apk add --update --no-cache --virtual .build-dependencies \
-        make git nodejs && \
-    make -C /go/src/github.com/cloudflare/unsee VERSION="${VERSION:-dev}" && \
-    mv /go/src/github.com/cloudflare/unsee/unsee /bin/unsee && \
-    rm -fr /go/src && \
-    apk del .build-dependencies
-
+FROM gcr.io/distroless/base
```
Comment (mattbostock) [inline Dockerfile]: What's the advantage of this over using the 'scratch' image and adding ca-certificates as necessary e.g. https://blog.codeship.com/building-minimal-docker-containers-for-go-applications/? It would be good to explain the rationale in your commit message.

Comment (prymitive) [inline Dockerfile]: gcr.io/distroless/base already is a minimal docker image with ca certs included
why maintain our own?

Comment (mattbostock) [inline Dockerfile]: What else does it pull in? Can we include that rationale in the commit message so that future contributors understand why we chose distroless?

File: Makefile (hunk 1 above)
Comment (prymitive) [inline Makefile]: We're no longer keeping resulting binary in the image it was build, so we need a more static build.
But it's likely that `CGO_ENABLED=0` is only needed really, @terinjokes ?

File: Dockerfile (hunk 2 above)
Comment (prymitive) [inline Dockerfile]: I already updated commit message, ca-cert is all we should need
see https://github.com/GoogleCloudPlatform/distroless

Comment (mattbostock) [inline Dockerfile]: I was struggling to find info as to what the distroless base image contains, but this covers it:

https://github.com/GoogleCloudPlatform/distroless/blob/master/base/README.md

We don't need glibc or libssl if the binary is statically compiled.

The code to use `scratch` would look roughly like:

```
# Build stage
FROM golang:1.8 as build
WORKDIR /go/src/github.com/foo/bar
RUN apt-get update
RUN apt-get upgrade -y ca-certificates
RUN apt-get install -y git make
COPY . /go/src/github.com/foo/bar
RUN make # or similar

# Main stage
FROM scratch
COPY --from=build /etc/ssl/certs /etc/ssl/certs
COPY --from=build /go/bin/example /
ENTRYPOINT ["/example"]
```

...which gives you an image containing only the statically-compiled binary and the CA certificates copied from Debian.

I'm not suggesting you change it, I think using distroless is fine.

File: Makefile (hunk 1 above)
Comment (mattbostock) [inline Makefile]: `CGO_ENABLED=0` will give you a statically-compiled Go binary.
